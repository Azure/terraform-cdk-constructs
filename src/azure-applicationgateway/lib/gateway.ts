import { ApplicationGateway } from "@cdktf/provider-azurerm/lib/application-gateway";
import * as azapgw from "@cdktf/provider-azurerm/lib/application-gateway";
import { KeyVault } from "@cdktf/provider-azurerm/lib/key-vault";
import { KeyVaultAccessPolicyA } from "@cdktf/provider-azurerm/lib/key-vault-access-policy";
import { PublicIp } from "@cdktf/provider-azurerm/lib/public-ip";
import { ResourceGroup } from "@cdktf/provider-azurerm/lib/resource-group";
import { Subnet } from "@cdktf/provider-azurerm/lib/subnet";
import { UserAssignedIdentity } from "@cdktf/provider-azurerm/lib/user-assigned-identity";
import { Construct } from "constructs";
import * as vnet from "../../azure-virtualnetwork";
import { AzureResource } from "../../core-azure";

// Define the interface for Application Gateway properties
export interface IGatewayProps {
  /**
   * Optional public IP address for the frontend of the Application Gateway.
   */
  publicIpAddress?: PublicIp;

  /**
   * Optional private IP address for the frontend of the Application Gateway.
   */
  privateIpAddress?: string;

  /**
   * Allocation method for the private IP address (e.g., Static, Dynamic).
   */
  privateIpAddressAllocation?: string;

  /**
   * The name of the Application Gateway.
   */
  readonly name: string;

  /**
   * The location where the Application Gateway will be deployed (e.g., region).
   */
  readonly location: string;

  /**
   * An optional reference to the resource group in which to deploy the Application Gateway.
   * If not provided, the Application Gateway will be deployed in the default resource group.
   */
  readonly resourceGroup?: ResourceGroup;

  /**
   * The SKU tier of the Application Gateway (e.g., Standard, WAF).
   */
  readonly skuTier: string;

  /**
   * The size of the SKU for the Application Gateway.
   */
  readonly skuSize: string;

  /**
   * The capacity (instance count) of the Application Gateway.
   */
  readonly capacity: number;

  /**
   * The backend address pools for the Application Gateway.
   */
  readonly backendAddressPools: azapgw.ApplicationGatewayBackendAddressPool[];

  /**
   * The backend HTTP settings for the Application Gateway.
   */
  readonly backendHttpSettings: azapgw.ApplicationGatewayBackendHttpSettings[];

  /**
   * Optional frontend ports for the Application Gateway.
   */
  readonly frontendPorts?: azapgw.ApplicationGatewayFrontendPort[];

  /**
   * The HTTP listeners for the Application Gateway.
   */
  readonly httpListeners: azapgw.ApplicationGatewayHttpListener[];

  /**
   * The request routing rules for the Application Gateway.
   */
  readonly requestRoutingRules: azapgw.ApplicationGatewayRequestRoutingRule[];

  /**
   * Optional subnet for the Application Gateway.
   */
  readonly subnet?: Subnet;

  /**
   * Optional tenant ID for use with Key Vault, if applicable.
   */
  readonly tenantId?: string;

  /**
   * Flag to enable HTTP2.
   */
  readonly enableHttp2?: boolean;

  /**
   * Flag to enable FIPS-compliant algorithms.
   */
  readonly fipsEnabled?: boolean;

  /**
   * Optional ID of the firewall policy.
   */
  readonly firewallPolicyId?: string;

  /**
   * Flag to enforce association of the firewall policy.
   */
  readonly forceFirewallPolicyAssociation?: boolean;

  /**
   * Optional tags for the Application Gateway resource.
   */
  readonly tags?: { [key: string]: string };

  /**
   * Optional availability zones for the Application Gateway.
   */
  readonly zones?: string[];

  /**
   * Optional Key Vault resource for storing SSL certificates.
   */
  readonly keyVault?: KeyVault;

  /**
   * Optional authentication certificates for mutual authentication.
   */
  readonly authenticationCertificate?: azapgw.ApplicationGatewayAuthenticationCertificate[];

  /**
   * Optional autoscale configuration for dynamically adjusting the capacity of the Application Gateway.
   */
  readonly autoscaleConfiguration?: azapgw.ApplicationGatewayAutoscaleConfiguration;

  /**
   * Optional custom error configurations to specify custom error pages.
   */
  readonly customErrorConfiguration?: azapgw.ApplicationGatewayCustomErrorConfiguration[];

  /**
   * Optional identity for the Application Gateway, used for accessing other Azure resources.
   */
  readonly identity?: azapgw.ApplicationGatewayIdentity;

  /**
   * Optional configurations for enabling Private Link on the Application Gateway.
   */
  readonly privateLinkConfiguration?: azapgw.ApplicationGatewayPrivateLinkConfiguration[];

  /**
   * Optional probes for health checks of the backend HTTP settings.
   */
  readonly probe?: azapgw.ApplicationGatewayProbe[];

  /**
   * Optional configurations for redirect rules.
   */
  readonly redirectConfiguration?: azapgw.ApplicationGatewayRedirectConfiguration[];

  /**
   * Optional rewrite rule sets for modifying HTTP request and response headers and bodies.
   */
  readonly rewriteRuleSet?: azapgw.ApplicationGatewayRewriteRuleSet[];

  /**
   * Optional SSL certificates for enabling HTTPS on the Application Gateway.
   */
  readonly sslCertificate?: azapgw.ApplicationGatewaySslCertificate[];

  /**
   * Optional SSL policy configurations, defining the protocol and cipher suites used.
   */
  readonly sslPolicy?: azapgw.ApplicationGatewaySslPolicy;

  /**
   * Optional SSL profiles for managing SSL termination and policy settings.
   */
  readonly sslProfile?: azapgw.ApplicationGatewaySslProfile[];

  /**
   * Optional timeout settings for the Application Gateway resources.
   */
  readonly timeouts?: azapgw.ApplicationGatewayTimeouts;

  /**
   * Optional trusted client certificates for mutual authentication.
   */
  readonly trustedClientCertificate?: azapgw.ApplicationGatewayTrustedClientCertificate[];

  /**
   * Optional trusted root certificates for backend authentication.
   */
  readonly trustedRootCertificate?: azapgw.ApplicationGatewayTrustedRootCertificate[];

  /**
   * Optional URL path map for routing based on URL paths.
   */
  readonly urlPathMap?: azapgw.ApplicationGatewayUrlPathMap[];

  /**
   * Optional Web Application Firewall (WAF) configuration to provide enhanced security.
   */
  readonly wafConfiguration?: azapgw.ApplicationGatewayWafConfiguration;
}

// Define the class for Azure Application Gateway
export class Gateway extends AzureResource {
  public readonly props: IGatewayProps;
  public resourceGroup: ResourceGroup;
  public id: string;

  /**
   * Constructs a new Azure Application Gateway.
   *
   * @param scope - The scope in which to define this construct.
   * @param id - The ID of this construct.
   * @param props - The properties for configuring the Azure Application Gateway. The properties include:
   *                - `name`: Required. Unique name for the Application Gateway within Azure.
   *                - `location`: Required. Azure Region for deployment.
   *                - `resourceGroup`: Optional. Reference to the resource group for deployment.
   *                - `skuTier`: Required. SKU tier of the Application Gateway (e.g., Standard, WAF).
   *                - `skuSize`: Required. Size of the SKU for the Application Gateway.
   *                - `capacity`: Required. Capacity (instance count) of the Application Gateway.
   *                - `backendAddressPools`: Required. Backend address pools for the Application Gateway.
   *                - `backendHttpSettings`: Required. Backend HTTP settings for the Application Gateway.
   *                - `httpListeners`: Required. HTTP listeners for the Application Gateway.
   *                - `requestRoutingRules`: Required. Request routing rules for the Application Gateway.
   *                - `publicIpAddress`: Optional. Public IP address for the frontend.
   *                - `privateIpAddress`: Optional. Private IP address for the frontend.
   *                - `privateIpAddressAllocation`: Optional. Allocation method for the private IP (Static, Dynamic).
   *                - `frontendPorts`: Optional. Frontend ports for the Application Gateway.
   *                - `subnet`: Optional. Subnet for the Application Gateway.
   *                - `enableHttp2`: Optional. Flag to enable HTTP2.
   *                - `fipsEnabled`: Optional. Flag to enable FIPS-compliant algorithms.
   *                - `firewallPolicyId`: Optional. ID of the firewall policy.
   *                - `forceFirewallPolicyAssociation`: Optional. Flag to enforce association of the firewall policy.
   *                - `tags`: Optional. Tags for resource management.
   *                - Additional optional properties as described in `IGatewayProps` interface.
   *
   * Example usage:
   * ```typescript
   * new Gateway(this, 'appGateway1', {
   *   name: 'gatewayEast',
   *   resourceGroup: resourceGroup,
      location: "eastus",
      skuTier: "Standard_v2",
      skuSize: "Standard_v2",
      capacity: 2,
      publicIpAddress: publicIp,
      subnet: subnet,
      backendAddressPools: [
        { name: "backend-address-pool-1" },
        {
          name: "backend-address-pool-2",
          ipAddresses: ["10.1.0.4", "10.1.0.5", "10.1.0.6"],
        },
      ],
      httpListeners: [
        {
          name: "http-listener",
          frontendPortName: "80",
          frontendIpConfigurationName: "Public-frontend-ip-configuration",
          protocol: "Http",
        },
      ],
      backendHttpSettings: [
        {
          name: "backend-http-setting",
          port: 80,
          protocol: "Http",
          requestTimeout: 20,
          cookieBasedAffinity: "Disabled",
        },
      ],
      requestRoutingRules: [
        {
          name: "request-routing-rule-1",
          httpListenerName: "http-listener",
          priority: 1,
          backendAddressPoolName: "backend-address-pool-1",
          backendHttpSettingsName: "backend-http-setting",
          ruleType: "Basic",
        },
      ],
   * });
   * ```
   */

  constructor(scope: Construct, id: string, props: IGatewayProps) {
    super(scope, id);

    this.props = props;
    this.resourceGroup = this.setupResourceGroup(props);

    // Define the identity
    let identity;

    // Check if Azure Key Vault is used and no identity is provided
    if (props.keyVault && !props.identity) {
      // Create a managed identity and add it to identityIds
      const managedIdentity = new UserAssignedIdentity(
        this,
        "ManagedIdentity",
        {
          name: `mi-${props.name}`,
          resourceGroupName: this.resourceGroup.name,
          location: props.location,
        },
      );

      identity = {
        identityIds: [managedIdentity.id],
        type: "UserAssigned",
      };

      new KeyVaultAccessPolicyA(this, "policy", {
        keyVaultId: props.keyVault.id,
        tenantId: props.tenantId || "",
        objectId: managedIdentity.principalId,
        secretPermissions: ["Get", "List"],
      });
    }

    const defaults = {
      subnetId:
        props.subnet?.id ||
        new vnet.Network(this, "vnet", {
          resourceGroup: this.resourceGroup,
        }).subnets.default.id,
      identity: props.identity || identity,
    };

    // WAF configuration
    let wafConfiguration = props.wafConfiguration
      ? {
          enabled: props.wafConfiguration.enabled,
          firewallMode: props.wafConfiguration.firewallMode || "Detection",
          ruleSetVersion: props.wafConfiguration.ruleSetVersion || "3.0",
        }
      : undefined;

    // Dynamically create frontend IP configurations
    let frontendIpConfigs : azapgw.ApplicationGatewayFrontendIpConfiguration[] = [];

    // Public IP configuration
    if (props.publicIpAddress) {
      frontendIpConfigs.push({
        name: "Public-frontend-ip-configuration",
        publicIpAddressId: props.publicIpAddress.id,
      });
    }

    // Private IP configuration
    if (props.privateIpAddress || props.privateIpAddressAllocation) {
      frontendIpConfigs.push({
        name: "Private-frontend-ip-configuration",
        subnetId: defaults.subnetId,
        privateIpAddress: props.privateIpAddress,
        privateIpAddressAllocation: props.privateIpAddressAllocation,
      });
    }

    // If no frontend ports are provided, use default dummy frontend ip configuration
    if (frontendIpConfigs.length == 0) {
      frontendIpConfigs.push({
        name: "Dummy-frontend-ip-configuration",
      });
    }

    // Set default frontend ports if not provided
    const defaultFrontendPorts = [
      { name: "80", port: 80 },
      { name: "443", port: 443 },
    ];

    const frontendPorts =
      props.frontendPorts && props.frontendPorts.length > 0
        ? props.frontendPorts
        : defaultFrontendPorts;

    // Create the Application Gateway
    const apgw = new ApplicationGateway(this, "ApplicationGateway", {
      name: props.name,
      resourceGroupName: this.resourceGroup.name,
      location: props.location,
      sslCertificate: props.sslCertificate,
      sslPolicy: props.sslPolicy,
      sslProfile: props.sslProfile,
      authenticationCertificate: props.authenticationCertificate,
      autoscaleConfiguration: props.autoscaleConfiguration,
      customErrorConfiguration: props.customErrorConfiguration,
      redirectConfiguration: props.redirectConfiguration,
      rewriteRuleSet: props.rewriteRuleSet,
      privateLinkConfiguration: props.privateLinkConfiguration,
      wafConfiguration: wafConfiguration,
      sku: {
        name: props.skuSize,
        tier: props.skuTier,
        capacity: props.capacity,
      },
      gatewayIpConfiguration: [
        {
          subnetId: defaults.subnetId,
          name: `${props.name}-configuration`,
        },
      ],
      frontendPort: frontendPorts,
      frontendIpConfiguration: frontendIpConfigs,
      backendAddressPool: props.backendAddressPools,
      backendHttpSettings: props.backendHttpSettings,
      httpListener: props.httpListeners,
      urlPathMap: props.urlPathMap,
      trustedRootCertificate: props.trustedRootCertificate,
      requestRoutingRule: props.requestRoutingRules,
      probe: props.probe,
      identity: defaults.identity,
      zones: props.zones,
      tags: props.tags,
    });

    this.id = apgw.id;
  }
}
