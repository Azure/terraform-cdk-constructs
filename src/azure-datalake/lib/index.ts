export * from "./filesystem";
export * from "./path";
export * from "./datalake";
