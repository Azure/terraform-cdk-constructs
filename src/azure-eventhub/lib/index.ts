export * from "./authorization";
export * from "./eventhubcluster";
export * from "./consumer";
export * from "./instance";
export * from "./kusto-connection";
export * from "./namespace";
