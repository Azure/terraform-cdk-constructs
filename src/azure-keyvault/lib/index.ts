export * from './certificate'
export * from './key'
export * from './policy'
export * from './secret'
export * from './vault'

