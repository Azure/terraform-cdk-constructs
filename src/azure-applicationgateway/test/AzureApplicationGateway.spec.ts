import { AzurermProvider } from "@cdktf/provider-azurerm/lib/provider";
import { PublicIp } from "@cdktf/provider-azurerm/lib/public-ip";
import { ResourceGroup } from "@cdktf/provider-azurerm/lib/resource-group";
import { Testing, TerraformStack } from "cdktf";
import { exampleAzureApplicationGateway } from "./ExampleAzureApplicationGateway";
import "cdktf/lib/testing/adapters/jest";
import * as apgw from "..";
import * as util from "../../util/azureTenantIdHelpers";

describe("Application Gateway With Defaults", () => {
  let stack: TerraformStack;
  let fullSynthResult: any;

  beforeEach(() => {
    const app = Testing.app();
    stack = new TerraformStack(app, "test");

    new AzurermProvider(stack, "azureFeature", { features: {} });

    const resourceGroup = new ResourceGroup(stack, "MyResourceGroup", {
      name: "testrg",
      location: "eastus",
    });

    const publicIp = new PublicIp(stack, "publicIp", {
      name: "testip",
      location: "eastus",
      resourceGroupName: resourceGroup.name,
      allocationMethod: "Dynamic",
      sku: "Standard",
    });

    new apgw.Gateway(stack, "testAzureApplicationGatewayDefaults2", {
      name: "application-gateway",
      resourceGroup: resourceGroup,
      location: "eastus",
      skuTier: "WAF_v2",
      skuSize: "WAF_v2",
      capacity: 2,
      publicIpAddress: publicIp,
      backendAddressPools: [
        { name: "backend-address-pool-1" },
        {
          name: "backend-address-pool-2",
          ipAddresses: ["10.0.0.4", "10.0.0.5", "10.0.0.6"],
        },
      ],
      httpListeners: [
        {
          name: "http-listener",
          frontendPortName: "frontend-port",
          frontendIpConfigurationName: "Public",
          protocol: "Http",
        },
      ],
      backendHttpSettings: [
        {
          name: "backend-http-setting",
          port: 80,
          protocol: "Http",
          requestTimeout: 20,
          cookieBasedAffinity: "Disabled",
        },
      ],
      requestRoutingRules: [
        {
          name: "request-routing-rule-1",
          httpListenerName: "http-listener",
          backendAddressPoolName: "backend-address-pool-1",
          backendHttpSettingsName: "backend-http-setting",
          ruleType: "Basic",
        },
        {
          name: "request-routing-rule-2",
          httpListenerName: "http-listener",
          backendAddressPoolName: "backend-address-pool-2",
          backendHttpSettingsName: "backend-http-setting",
          ruleType: "Basic",
        },
      ],
    });

    fullSynthResult = Testing.fullSynth(stack); // Save the result for reuse
  });

  it("renders an Application Gateway with defaults and checks snapshot", () => {
    expect(Testing.synth(stack)).toMatchSnapshot(); // Compare the already prepared stack
  });

  it("check if the produced terraform configuration is valid", () => {
    expect(fullSynthResult).toBeValidTerraform(); // Use the saved result
  });

  it("check if this can be planned", () => {
    expect(fullSynthResult).toPlanSuccessfully(); // Use the saved result
  });
});

describe("Application Gateway Example", () => {
  it("renders the Application Gateway and checks snapshot", () => {
    // Need to remove the tenant_id from the snapshot as it will change wherever the test is run
    const output = Testing.synth(
      new exampleAzureApplicationGateway(
        Testing.app(),
        "testAzureApplicationGateway",
      ),
    );
    const myObject: Record<string, any> = JSON.parse(output);

    expect(util.removeTenantIdFromSnapshot(myObject)).toMatchSnapshot();
  });

  it("check if the produced terraform configuration is valid", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(
      Testing.fullSynth(
        new exampleAzureApplicationGateway(
          Testing.app(),
          "testAzureApplicationGateway",
        ),
      ),
    ).toBeValidTerraform();
  });

  it("check if this can be planned", () => {
    // We need to do a full synth to plan the terraform configuration
    expect(
      Testing.fullSynth(
        new exampleAzureApplicationGateway(
          Testing.app(),
          "testAzureApplicationGateway",
        ),
      ),
    ).toPlanSuccessfully();
  });
});
