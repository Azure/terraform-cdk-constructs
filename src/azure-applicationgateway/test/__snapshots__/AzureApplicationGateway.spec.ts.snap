// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application Gateway Example renders the Application Gateway and checks snapshot 1`] = `
{
  "data": {
    "azurerm_client_config": {
      "CurrentClientConfig": {},
    },
    "azurerm_monitor_diagnostic_categories": {
      "appgw_diagsettings_diagcategories_A1E0B03C": {
        "resource_id": "\${azurerm_application_gateway.appgw_ApplicationGateway_698001D9.id}",
      },
    },
  },
  "output": {
    "id": {
      "value": "\${azurerm_key_vault.keyvault_key_vault_53A151E3.id}",
    },
    "key_vault_name": {
      "value": "\${azurerm_key_vault.keyvault_key_vault_53A151E3.name}",
    },
    "resource_group_name": {
      "value": "\${azurerm_resource_group.rg.name}",
    },
  },
  "provider": {
    "azurerm": [
      {
        "features": {
          "resource_group": {
            "prevent_deletion_if_contains_resources": false,
          },
        },
      },
    ],
  },
  "resource": {
    "azurerm_application_gateway": {
      "appgw_ApplicationGateway_698001D9": {
        "backend_address_pool": [
          {
            "name": "backend-address-pool-1",
          },
          {
            "ip_addresses": [
              "10.1.0.4",
              "10.1.0.5",
              "10.1.0.6",
            ],
            "name": "backend-address-pool-2",
          },
        ],
        "backend_http_settings": [
          {
            "cookie_based_affinity": "Disabled",
            "name": "backend-http-setting",
            "port": 80,
            "protocol": "Http",
            "request_timeout": 20,
          },
        ],
        "frontend_ip_configuration": [
          {
            "name": "Public-frontend-ip-configuration",
            "public_ip_address_id": "\${azurerm_public_ip.publicIp.id}",
          },
        ],
        "frontend_port": [
          {
            "name": "80",
            "port": 80,
          },
          {
            "name": "443",
            "port": 443,
          },
        ],
        "gateway_ip_configuration": [
          {
            "name": "apgw-\${var.name}-configuration",
            "subnet_id": "\${azurerm_subnet.subnet.id}",
          },
        ],
        "http_listener": [
          {
            "frontend_ip_configuration_name": "Public-frontend-ip-configuration",
            "frontend_port_name": "80",
            "name": "http-listener",
            "protocol": "Http",
          },
        ],
        "location": "eastus",
        "name": "apgw-\${var.name}",
        "request_routing_rule": [
          {
            "backend_address_pool_name": "backend-address-pool-1",
            "backend_http_settings_name": "backend-http-setting",
            "http_listener_name": "http-listener",
            "name": "request-routing-rule-1",
            "priority": 1,
            "rule_type": "Basic",
          },
        ],
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "sku": {
          "capacity": 2,
          "name": "Standard_v2",
          "tier": "Standard_v2",
        },
      },
      "appgw_waf_ApplicationGateway_E57FD855": {
        "backend_address_pool": [
          {
            "name": "backend-address-pool-1",
          },
          {
            "ip_addresses": [
              "10.1.0.4",
              "10.1.0.5",
              "10.1.0.6",
            ],
            "name": "backend-address-pool-2",
          },
        ],
        "backend_http_settings": [
          {
            "cookie_based_affinity": "Disabled",
            "name": "backend-http-setting",
            "port": 80,
            "protocol": "Http",
            "request_timeout": 20,
          },
        ],
        "frontend_ip_configuration": [
          {
            "name": "Public-frontend-ip-configuration",
            "public_ip_address_id": "\${azurerm_public_ip.publicIpwaf.id}",
          },
        ],
        "frontend_port": [
          {
            "name": "80",
            "port": 80,
          },
          {
            "name": "443",
            "port": 443,
          },
        ],
        "gateway_ip_configuration": [
          {
            "name": "apgw-\${var.name}waf-configuration",
            "subnet_id": "\${azurerm_subnet.subnet.id}",
          },
        ],
        "http_listener": [
          {
            "frontend_ip_configuration_name": "Public-frontend-ip-configuration",
            "frontend_port_name": "443",
            "name": "http-listener",
            "protocol": "Https",
            "ssl_certificate_name": "internal.contoso.com",
          },
        ],
        "identity": {
          "identity_ids": [
            "\${azurerm_user_assigned_identity.appgw_waf_ManagedIdentity_5D2DB2C2.id}",
          ],
          "type": "UserAssigned",
        },
        "location": "eastus",
        "name": "apgw-\${var.name}waf",
        "probe": [
          {
            "host": "test",
            "interval": 3,
            "name": "probe-8080",
            "path": "/status",
            "port": 8080,
            "protocol": "Http",
            "timeout": 15,
            "unhealthy_threshold": 3,
          },
        ],
        "redirect_configuration": [
          {
            "include_path": true,
            "include_query_string": true,
            "name": "http-redirect",
            "redirect_type": "Permanent",
            "target_listener_name": "http-listener",
          },
        ],
        "request_routing_rule": [
          {
            "backend_address_pool_name": "backend-address-pool-1",
            "backend_http_settings_name": "backend-http-setting",
            "http_listener_name": "http-listener",
            "name": "request-routing-rule-1",
            "priority": 1,
            "rule_type": "Basic",
          },
        ],
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "sku": {
          "capacity": 2,
          "name": "WAF_v2",
          "tier": "WAF_v2",
        },
        "ssl_certificate": [
          {
            "key_vault_secret_id": "\${azurerm_key_vault_certificate.keyvault_ssl-cert_AzureKeyVaultCertificate_9B2476CE.secret_id}",
            "name": "internal.contoso.com",
          },
        ],
        "url_path_map": [
          {
            "default_redirect_configuration_name": "http-redirect",
            "name": "url-path-map",
            "path_rule": [
              {
                "name": "path-rule-01",
                "paths": [
                  "/helloworld",
                ],
                "redirect_configuration_name": "http-redirect",
              },
            ],
          },
        ],
        "waf_configuration": {
          "enabled": true,
          "firewall_mode": "Detection",
          "rule_set_version": "3.0",
        },
      },
    },
    "azurerm_key_vault": {
      "keyvault_key_vault_53A151E3": {
        "location": "eastus",
        "name": "kv-\${var.name}",
        "purge_protection_enabled": true,
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "sku_name": "standard",
        "soft_delete_retention_days": 90,
      },
    },
    "azurerm_key_vault_access_policy": {
      "appgw_waf_policy_F1E21F34": {
        "key_vault_id": "\${azurerm_key_vault.keyvault_key_vault_53A151E3.id}",
        "object_id": "\${azurerm_user_assigned_identity.appgw_waf_ManagedIdentity_5D2DB2C2.principal_id}",
        "secret_permissions": [
          "Get",
          "List",
        ],
      },
      "keyvault_kv_custom_policy_access_TfTokenTOKEN75_policy_450D91E5": {
        "certificate_permissions": [
          "Get",
          "List",
          "Create",
          "Delete",
          "GetIssuers",
          "ManageIssuers",
        ],
        "key_permissions": [
          "Backup",
          "Create",
          "Decrypt",
          "Delete",
          "Encrypt",
          "Get",
          "Import",
          "List",
          "Purge",
          "Recover",
          "Restore",
          "Sign",
          "UnwrapKey",
          "Update",
          "Verify",
          "WrapKey",
          "Release",
          "Rotate",
          "GetRotationPolicy",
          "SetRotationPolicy",
        ],
        "key_vault_id": "\${azurerm_key_vault.keyvault_key_vault_53A151E3.id}",
        "object_id": "\${data.azurerm_client_config.CurrentClientConfig.object_id}",
        "secret_permissions": [
          "Get",
          "List",
          "Set",
          "Delete",
        ],
      },
    },
    "azurerm_key_vault_certificate": {
      "keyvault_ssl-cert_AzureKeyVaultCertificate_9B2476CE": {
        "certificate_policy": {
          "issuer_parameters": {
            "name": "Self",
          },
          "key_properties": {
            "exportable": true,
            "key_size": 2048,
            "key_type": "RSA",
            "reuse_key": true,
          },
          "lifetime_action": [
            {
              "action": {
                "action_type": "AutoRenew",
              },
              "trigger": {
                "days_before_expiry": 45,
              },
            },
          ],
          "secret_properties": {
            "content_type": "application/x-pkcs12",
          },
          "x509_certificate_properties": {
            "extended_key_usage": [
              "1.3.6.1.5.5.7.3.1",
            ],
            "key_usage": [
              "cRLSign",
              "dataEncipherment",
              "digitalSignature",
              "keyAgreement",
              "keyCertSign",
              "keyEncipherment",
            ],
            "subject": "CN=contoso.com",
            "subject_alternative_names": {
              "dns_names": [
                "internal.contoso.com",
              ],
            },
            "validity_in_months": 12,
          },
        },
        "depends_on": [
          "azurerm_key_vault_access_policy.keyvault_kv_custom_policy_access_TfTokenTOKEN75_policy_450D91E5",
        ],
        "key_vault_id": "\${azurerm_key_vault.keyvault_key_vault_53A151E3.id}",
        "name": "ssl-cert",
      },
    },
    "azurerm_log_analytics_workspace": {
      "log_analytics": {
        "location": "eastus",
        "name": "la-\${var.name}",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
      },
    },
    "azurerm_monitor_diagnostic_setting": {
      "appgw_diagsettings_8B8E708F": {
        "dynamic": {
          "enabled_log": {
            "content": {
              "category": "\${enabled_log.value}",
            },
            "for_each": "\${tolist(data.azurerm_monitor_diagnostic_categories.appgw_diagsettings_diagcategories_A1E0B03C.log_category_types)}",
          },
          "metric": {
            "content": {
              "category": "\${metric.value}",
            },
            "for_each": "\${tolist(data.azurerm_monitor_diagnostic_categories.appgw_diagsettings_diagcategories_A1E0B03C.metrics)}",
          },
        },
        "log_analytics_workspace_id": "\${azurerm_log_analytics_workspace.log_analytics.id}",
        "name": "diagsettings",
        "target_resource_id": "\${azurerm_application_gateway.appgw_ApplicationGateway_698001D9.id}",
      },
    },
    "azurerm_public_ip": {
      "publicIp": {
        "allocation_method": "Static",
        "location": "eastus",
        "name": "testip",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "sku": "Standard",
      },
      "publicIpwaf": {
        "allocation_method": "Static",
        "location": "eastus",
        "name": "testipwaf",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "sku": "Standard",
      },
    },
    "azurerm_resource_group": {
      "rg": {
        "lifecycle": {
          "ignore_changes": [
            "tags",
          ],
        },
        "location": "eastus",
        "name": "rg-\${var.name}",
      },
    },
    "azurerm_role_assignment": {
      "appgw_TfTokenTOKEN96Contributor_role_68A30423": {
        "principal_id": "\${data.azurerm_client_config.CurrentClientConfig.object_id}",
        "role_definition_name": "Contributor",
        "scope": "\${azurerm_application_gateway.appgw_ApplicationGateway_698001D9.id}",
      },
    },
    "azurerm_subnet": {
      "subnet": {
        "address_prefixes": [
          "10.0.0.0/24",
        ],
        "name": "subnet-\${var.name}",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
        "virtual_network_name": "\${azurerm_virtual_network.vnet.name}",
      },
    },
    "azurerm_user_assigned_identity": {
      "appgw_waf_ManagedIdentity_5D2DB2C2": {
        "location": "eastus",
        "name": "mi-apgw-\${var.name}waf",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
      },
    },
    "azurerm_virtual_network": {
      "vnet": {
        "address_space": [
          "10.0.0.0/24",
        ],
        "location": "eastus",
        "name": "vnet-\${var.name}",
        "resource_group_name": "\${azurerm_resource_group.rg.name}",
      },
    },
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.70.0",
      },
    },
  },
  "variable": {
    "name": {
      "default": "test",
      "description": "System name used to randomize the resources",
      "type": "string",
    },
  },
}
`;

exports[`Application Gateway With Defaults renders an Application Gateway with defaults and checks snapshot 1`] = `
"{
  "provider": {
    "azurerm": [
      {
        "features": {
        }
      }
    ]
  },
  "resource": {
    "azurerm_application_gateway": {
      "testAzureApplicationGatewayDefaults2_ApplicationGateway_04B13FBA": {
        "backend_address_pool": [
          {
            "name": "backend-address-pool-1"
          },
          {
            "ip_addresses": [
              "10.0.0.4",
              "10.0.0.5",
              "10.0.0.6"
            ],
            "name": "backend-address-pool-2"
          }
        ],
        "backend_http_settings": [
          {
            "cookie_based_affinity": "Disabled",
            "name": "backend-http-setting",
            "port": 80,
            "protocol": "Http",
            "request_timeout": 20
          }
        ],
        "frontend_ip_configuration": [
          {
            "name": "Public-frontend-ip-configuration",
            "public_ip_address_id": "\${azurerm_public_ip.publicIp.id}"
          }
        ],
        "frontend_port": [
          {
            "name": "80",
            "port": 80
          },
          {
            "name": "443",
            "port": 443
          }
        ],
        "gateway_ip_configuration": [
          {
            "name": "application-gateway-configuration",
            "subnet_id": "\${azurerm_subnet.testAzureApplicationGatewayDefaults2_vnet_default_94DB58C1.id}"
          }
        ],
        "http_listener": [
          {
            "frontend_ip_configuration_name": "Public",
            "frontend_port_name": "frontend-port",
            "name": "http-listener",
            "protocol": "Http"
          }
        ],
        "location": "eastus",
        "name": "application-gateway",
        "request_routing_rule": [
          {
            "backend_address_pool_name": "backend-address-pool-1",
            "backend_http_settings_name": "backend-http-setting",
            "http_listener_name": "http-listener",
            "name": "request-routing-rule-1",
            "rule_type": "Basic"
          },
          {
            "backend_address_pool_name": "backend-address-pool-2",
            "backend_http_settings_name": "backend-http-setting",
            "http_listener_name": "http-listener",
            "name": "request-routing-rule-2",
            "rule_type": "Basic"
          }
        ],
        "resource_group_name": "\${azurerm_resource_group.MyResourceGroup.name}",
        "sku": {
          "capacity": 2,
          "name": "WAF_v2",
          "tier": "WAF_v2"
        }
      }
    },
    "azurerm_public_ip": {
      "publicIp": {
        "allocation_method": "Dynamic",
        "location": "eastus",
        "name": "testip",
        "resource_group_name": "\${azurerm_resource_group.MyResourceGroup.name}",
        "sku": "Standard"
      }
    },
    "azurerm_resource_group": {
      "MyResourceGroup": {
        "location": "eastus",
        "name": "testrg"
      }
    },
    "azurerm_subnet": {
      "testAzureApplicationGatewayDefaults2_vnet_default_94DB58C1": {
        "address_prefixes": [
          "10.1.0.0/24"
        ],
        "name": "default",
        "resource_group_name": "\${azurerm_resource_group.MyResourceGroup.name}",
        "virtual_network_name": "\${azurerm_virtual_network.testAzureApplicationGatewayDefaults2_vnet_33CC9CBF.name}"
      }
    },
    "azurerm_virtual_network": {
      "testAzureApplicationGatewayDefaults2_vnet_33CC9CBF": {
        "address_space": [
          "10.0.0.0/16"
        ],
        "location": "eastus",
        "name": "vnet-test",
        "resource_group_name": "\${azurerm_resource_group.MyResourceGroup.name}"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "azurerm",
        "version": "3.70.0"
      }
    }
  }
}"
`;
